lib/fshash/metadata.go

```diff
@@ -0,0 +1,118 @@
+package fshash
+
+import (
+	"archive/tar"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/spacemonkeygo/errors"
+	"github.com/ugorji/go/codec"
+	"polydawn.net/repeatr/def"
+)
+
+type Metadata tar.Header
+
+func ReadMetadata(path string, optional ...os.FileInfo) Metadata {
+	var fi os.FileInfo
+	var err error
+	if len(optional) > 0 {
+		fi = optional[0]
+	} else {
+		fi, err = os.Lstat(path)
+		if err != nil {
+			// also consider ENOEXIST a problem; this function is mostly
+			// used in testing where we really expect that path to exist.
+			panic(errors.IOError.Wrap(err))
+		}
+	}
+	// readlink needs the file path again  ヽ(´ー｀)ノ
+	var link string
+	if fi.Mode()&os.ModeSymlink == os.ModeSymlink {
+		if link, err = os.Readlink(path); err != nil {
+			panic(errors.IOError.Wrap(err))
+		}
+	}
+	hdr, err := tar.FileInfoHeader(fi, link)
+	if err != nil {
+		panic(errors.IOError.Wrap(err))
+	}
+	// ctimes are uncontrollable, pave them (╯°□°）╯︵ ┻━┻
+	// atimes mutate on read, pave them
+	hdr.ChangeTime = def.Somewhen
+	hdr.AccessTime = def.Somewhen
+	return Metadata(*hdr)
+}
+
+func (m Metadata) Marshal(out io.Writer) {
+	// Encodes the metadata as a CBOR map -- deterministically; the output is appropriate to feed to a hash and expect consistency.
+	// We follow the rfc7049 section 3.9 description of "canonical CBOR": namely, map keys are here entered consistently, and in sorted order.
+	// Except when maps are representing a struct; then it's deterministic order, but specified by (fairly arbitrary) hardcoded choices.
+	// This doesn't implement `BinaryMarshaller` because we A: don't care and B: are invariably writing to another stream anyway.
+	// Note that if your writer ever returns an error, the codec library will panic with exactly that.  Yes, including `io.EOF`.
```

Comment:
May as well bump this docstring up above the declaration for godoc

Current output:

```
...
func ReadMetadata(path string, optional ...os.FileInfo) Metadata

func (m Metadata) Marshal(out io.Writer)

type Record struct {
    Metadata    Metadata `json:"m"`
    ContentHash []byte   `json:"h"`
}
...
```